### ISSUE TITLE
Mask within which to calculate loss

### ISSUE THREAD

**Issue by tknapen** [rohitrango/FireANTs#17]
One option many packages like ANTs have is to calculate the loss not over the entire image, but to allow the application of a mask in either fixed or moving space. A slightly more advanced use case, this, but it allows for quite handy fine-tuning of registration pipelines.
With this functionality added, I could rewrite my entire pipelines into fireants...

**Comment by rohitrango** [comment]
this is a feature not standardized across FireANTs yet, but what I've found is you can mask the images themselves before registering them. A simple workaround fix for now is 

```python
image = Image.load_file('/path/to/image')
mask = Image.load_file('/path/to/mask')
image.array = image.array * mask.array
del mask

# ... subsequent registration ...
``` 

This approach is more scalable than modifying every loss function, etc. but let me know what your existing pipelines are, so I can look at them.

**Comment by gdevenyi** [comment]
Brain extraction is not equivalent to masking because the smoothing operations will be between zero and the data instead of the "important data" and the "not important data"

**Comment by rohitrango** [comment]
what do you mean by "brain extraction" - like a [probability map](https://dpaniukov.github.io/2016/06/06/brain-extraction-with-ants.html)?  

I agree that masking the images (with possibly a non-zero value) *is* different from masking the loss itself. However, there are multiple considerations when masking losses and all of them have their own merits (and demerits):

1. If both images come with masks, how is the loss calculated with the provided masks? Is the loss computed with the intersections of the fixed and moved masks, or their union? Is there any principled answer as to why? Is the (lack of) mask overlap penalized separately?
	1.  if intersection of masks is used, then registration can discard too much data, especially early in training when alignment is poor. In the extreme case, the intersection might degenerate to near-zero support if masks don't overlap initially (this is a rare case since an initial pre-alignment is done).
	2. if union of masks is used, then irrelevant regions like artifacts from outside the ROI is included in the loss, which (a) dilutes the loss, and (b) the irrelevant regions might lead to the warp field in those regions stuck to a local minima.
2. If only one mask is provided (say moving mask), is the loss computed as a sum or mean over the regions? If the sum is used, the algorithm can trivially shrink the mask out of regions with high loss, instead of attempting to align the relevant anatomy in those high-loss regions. If the average is used for calculating the loss, then there is an extra $\frac{1}{\sum M(x)}$ term that is biased against small masks - if the region of interest is indeed small, then this term will make the loss higher than if the mask were dilated to include spurious homogenous regions that don't contribute much to the term $l(f(x), m(\varphi(x)))$  but simultaneously decrease the term $\frac{1}{\sum M(x)}$ due to the dilation of the ROI.  Moreover, for loss functions that are reducible to a sum or mean (mutual information, dice score), there are undefined effects of regions going in and out of the mask. 
	1. the most "stable" condition I can think of - is if only the fixed mask were provided, because the $\frac{1}{\sum M(x)}$ term does not change. Even then, regions moving in and out of the mask can lead to spurious matching and/or undefined behavior of gradient descent algorithms.


**Q. What happens when images are masked instead?**

If both images are masked with an appropriate background value (beyond the range of image intensities), then any overlap of background with background is not penalized, and any overlap of background region and non background region between the two images is penalized heavily.  Within regions that are both non-background, the loss is determined by the actual registration quality. ROIs with high loss (unregistered anatomy) cannot move out of the "masked region" because then those regions will have to overlap with background from the other image. So the only way for the ROIs with high loss is to find their correspondence. Moreover, loss functions like mutual information and/or dice scores can be applied with masked images as well.

In my experience (and most modern methods), the images/ROIs are masked first before the registration -  see [ANTs in EMPIRE10](https://drive.google.com/drive/folders/1cB9tuH0JK4GmYe8ce_PmLrmoUKHqdvYx) where the lung regions are masked and normalized appropriately before registration, [OASIS dataset](https://learn2reg.grand-challenge.org/Datasets/) and other brain datasets where skull-stripping is done routinely before registration, [convexAdam](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=10681158) where MIND features are computed after masking the images for modalities like lung and abdomen. I had initially played around with the idea of just a single mask (you will see code remnants in `affine.py`) but quickly reverted it because there are (a) too many undefined behaviors that can occur which the user will find hard to debug, (b) is not "cleanly compatible" with loss functions like mutual information, dice. One of the strong selling points of FireANTs is that you can write your own modality specific losses without learning ITK template C++ code and managing superbuilds, and masking within the loss complicates things.  

Unfortunately, I strongly believe there is no correct answer to this, and its a philosophical / implementation preference. I want to keep FireANTs fairly accurate, fast, accessible and usable across a broad range of tasks. So far I have noticed that for most applications masking the images suffices for downstream analysis. If you only have one mask, you may want to rethink about the anatomy you are trying to register for which you do not have the mask.
